README.md
python-pointdns
===========

pointhq.com API client.

This module provides easy access to point zone & record management. For information about the services offered on Point see [the website](http://pointhq.com)
This module is based on a fork of [Richard Gray's](https://github.com/vortura/pointhq) pointhq module which is based on [Mike Yumatov's](https://github.com/yumike/pointhq) pointhq module
with the following additional modifications:

- Added support for Python3.x
- Retrieve zones by groups
- Record type validation

## Authentication

To access your Point account, you'll need to define your username & apitoken. The username is your email address and the apitoken is the API token which, can be found in My Account tab.

## Installation

Install ``pointdns`` with pip:

    $ pip install pointdns

or directly from GitHub

    $ pip install git+https://github.com/cognetoapps/python-pointdns.git

It will also install `requests` library.

## Usage example

* Create new `pointdns.Point` object:
```python
from pointdns import Point
point = Point(username='john@example.com', apitoken='secret-key')
```
* Play with zones::
```python
zones = point.zones.retrieve()
new_zone = point.zones.create(name='example.com')
```
```python
zone = point.zones(1).retrieve()
point.zones(1).update(group='Clients')
zones_group_clients = point.zones.retrieve(group='Clients')
point.zones(1).delete()
```
* Play with zone records::
```python
zone_records = point.zones(1).records.retrieve()
new_record = point.zones(1).records.create(
                name='example.com.', data='123.45.67.89', record_type='A')
```
```python
zone_record = point.zones(1).records(1).retrieve
point.zones(1).records(1).update(data='234.56.78.90')
point.zones(1).records(1).delete()
```
## Contributing

Feel free to fork, send pull requests or report bugs and issues on github.


HISTORY.rst
History
=======

0.1 (2014-04-22)
----------------

- Initial release


0.1.1(2015-12-14)
----------------

- Fix updates


tests/test_zones.py
from __future__ import with_statement

import unittest2
from mock import Mock

from pointdns import Point
from pointdns import exceptions
from pointdns.helpers import Response


class ZoneTests(unittest2.TestCase):

    def setUp(self):
        self.auth = ('john@example.com', 'secret-key')
        self.request = Mock()
        self.point = Point(self.auth[0], self.auth[1], self.request)

    def test_collection_retrieve(self):
        self.request.return_value = Response(200, '[{"zone": {"id": 1}}]')
        zone = self.point.zones.retrieve()
        self.request.assert_called_once_with('get', '/zones', self.auth, None, {})
        self.assertEqual(zone, [{'zone': {'id': 1}}])

    def test_collection_retrieve_by_groups(self):
        self.request.return_value = Response(200, '[{"zone": {"id": 1}}]')
        zone = self.point.zones.retrieve(group='Clients')
        self.request.assert_called_once_with('get', '/zones', self.auth, None, {'group': 'Clients'})
        self.assertEqual(zone, [{'zone': {'id': 1}}])

    def test_collection_retrieve_failure(self):
        self.request.return_value = Response(201)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones.retrieve()

    def test_collection_create(self):
        self.request.return_value = Response(201, '{"zone": {"id": 1}}')
        zone = self.point.zones.create(name='example.com')
        self.request.assert_called_once_with('post', '/zones', self.auth, {
            'zone': {
                'name': 'example.com',
            },
        }, None)
        self.assertEqual(zone, {'zone': {'id': 1}})

    def test_collection_create_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones.create(name='example.com')

    def test_collection_create_unknown_failure(self):
        self.request.return_value = Response(200)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones.create(name='example.com')

    def test_member_retrieve(self):
        self.request.return_value = Response(200, '{"zone": {"id": 1}}')
        zone = self.point.zones(1).retrieve()
        self.request.assert_called_once_with('get', '/zones/1', self.auth, None, None)
        self.assertEqual(zone, {'zone': {'id': 1}})

    def test_member_retrieve_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.NotFoundError):
            self.point.zones(1).retrieve()

    def test_member_retrieve_unknown_failure(self):
        self.request.return_value = Response(201)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).retrieve()

    def test_member_update(self):
        self.request.return_value = Response(202)
        self.assertIsNone(self.point.zones(1).update(group='clients'))
        self.request.assert_called_once_with('put', '/zones/1', self.auth, {
            'zone': {
                'group': 'clients',
            }
        }, None)

    def test_member_update_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones(1).update(group='clients')

    def test_member_update_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).update(group='clients')

    def test_member_delete(self):
        self.request.return_value = Response(202)
        self.assertIsNone(self.point.zones(1).delete())
        self.request.assert_called_once_with('delete', '/zones/1', self.auth, None, None)

    def test_member_delete_failure(self):
        self.request.return_value = Response(409, '["error message"]')
        with self.assertRaises(exceptions.ConflictError):
            self.point.zones(1).delete()

    def test_member_delete_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).delete()


tests/test_request.py
from pointdns.helpers import request
import unittest2
from httmock import urlmatch, HTTMock


class RequestTests(unittest2.TestCase):

    def test_https_post_request(self):

        @urlmatch(netloc=r'pointhq\.com', scheme='https',
                  method='post', path='/')
        def response_content(url, request):
            return {'status_code': 200,
                    'content': b'OK'}

        with HTTMock(response_content):
            r = request('post', '/', ('john', 'secret-key'), scheme='https')
            self.assertTrue(r.status == 200)
            self.assertTrue(r.content == 'OK')

    def test_http_get_request(self):

        @urlmatch(netloc=r'pointhq\.com', scheme='http',
                  method='get', path='/')
        def response_content(url, request):
            return {'status_code': 200,
                    'content': b'OK'}

        with HTTMock(response_content):
            r = request('get', '/', ('john', 'secret-key'), scheme='http')
            self.assertTrue(r.status == 200)
            self.assertTrue(r.content == 'OK')

    def test_http_put_request(self):

        @urlmatch(netloc=r'pointhq\.com', scheme='http',
                  method='put', path='/')
        def response_content(url, request):
            return {'status_code': 200,
                    'content': b'OK'}

        with HTTMock(response_content):
            r = request('put', '/', ('john', 'secret-key'), scheme='http')
            self.assertTrue(r.status == 200)
            self.assertTrue(r.content == 'OK')

    def test_https_put_request(self):

        @urlmatch(netloc=r'pointhq\.com', scheme='https',
                  method='put', path='/')
        def response_content(url, request):
            return {'status_code': 200,
                    'content': b'OK'}

        with HTTMock(response_content):
            r = request('put', '/', ('john', 'secret-key'), scheme='https')
            self.assertTrue(r.status == 200)
            self.assertTrue(r.content == 'OK')


tests/test_zone_records.py
from __future__ import with_statement

import unittest2
from mock import Mock

from pointdns import Point
from pointdns import exceptions
from pointdns.helpers import Response


class ZoneRecordTests(unittest2.TestCase):

    def setUp(self):
        self.auth = ('john@example.com', 'secret-key')
        self.request = Mock()
        self.point = Point(self.auth[0], self.auth[1], self.request)

    def test_collection_retrieve(self):
        self.request.return_value = Response(200, '[{"zone_record": {"id": 1}}]')
        zone_record = self.point.zones(1).records.retrieve()
        self.request.assert_called_once_with('get', '/zones/1/records', self.auth, None, {})
        self.assertEqual(zone_record, [{'zone_record': {'id': 1}}])

    def test_collection_retrieve_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records.retrieve()

    def test_collection_create(self):
        self.request.return_value = Response(201, '{"zone_record": {"id": 1}}')
        zone_record = self.point.zones(1).records.create(name='example.com.', data='123.45.67.89')
        self.request.assert_called_once_with('post', '/zones/1/records', self.auth, {
            'zone_record': {
                'name': 'example.com.',
                'data': '123.45.67.89',
            }
        }, None)
        self.assertEqual(zone_record, {'zone_record': {'id': 1}})

    def test_collection_create_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones(1).records.create(name='example.com.')

    def test_collection_create_unknown_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records.create(name='example.com.')

    def test_member_retrieve(self):
        self.request.return_value = Response(200, '{"zone_record": {"id": 1}}')
        zone_record = self.point.zones(1).records(1).retrieve()
        self.request.assert_called_once_with('get', '/zones/1/records/1', self.auth, None, None)
        self.assertEqual(zone_record, {'zone_record': {'id': 1}})

    def test_member_retrieve_failure(self):
        self.request.return_value = Response(404)
        with self.assertRaises(exceptions.NotFoundError):
            self.point.zones(1).records(1).retrieve()

    def test_member_retrieve_unknown_failure(self):
        self.request.return_value = Response(500)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records(1).retrieve()

    def test_member_update(self):
        self.request.return_value = Response(202)
        self.assertIsNone(self.point.zones(1).records(1).update(data='123.45.67.89'))
        self.request.assert_called_once_with('put', '/zones/1/records/1', self.auth, {
            'zone_record': {
                'data': '123.45.67.89',
            }
        }, None)

    def test_member_update_failure(self):
        self.request.return_value = Response(422, '["error message"]')
        with self.assertRaises(exceptions.UnprocessableEntityError):
            self.point.zones(1).records(1).update(data='123.45.67.89')

    def test_member_update_unknown_failure(self):
        self.request.return_value = Response(500)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records(1).update(data='123.45.67.89')

    def test_member_delete(self):
        self.request.return_value = Response(202)
        self.assertIsNone(self.point.zones(1).records(1).delete())
        self.request.assert_called_once_with('delete', '/zones/1/records/1', self.auth, None, None)

    def test_member_delete_failure(self):
        self.request.return_value = Response(409)
        with self.assertRaises(exceptions.ConflictError):
            self.point.zones(1).records(1).delete()

    def test_member_delete_unknown_failure(self):
        self.request.return_value = Response(500)
        with self.assertRaises(exceptions.UnknownError):
            self.point.zones(1).records(1).delete()


pointdns/api.py
from . import base
from . import helpers


class Point(object):

    def __init__(self, username, apitoken, request_func=helpers.request):
        self.auth = (username, apitoken)
        self.request_func = request_func

    @property
    def zones(self):
        return ZoneCollection(request_func=self.request)

    def request(self, method, url, data=None, query_data=None):
        return self.request_func(method, url, self.auth, data, query_data)


class Zone(base.BaseMember):

    name = 'zone'
    path_pattern = '/zones/%(id)s'

    @property
    def records(self):
        return ZoneRecordCollection(
            zone_id=self.kwargs['id'],
            request_func=self.kwargs['request_func'],
        )


class ZoneCollection(base.BaseCollection):

    name = 'zone'
    path_pattern = '/zones'
    member_class = Zone


class ZoneRecord(base.BaseMember):

    name = 'zone_record'
    path_pattern = '/zones/%(zone_id)s/records/%(id)s'


class ZoneRecordCollection(base.BaseCollection):

    name = 'zone_record'
    path_pattern = '/zones/%(zone_id)s/records'
    member_class = ZoneRecord


pointdns/base.py
import json

from functools import wraps
from . import exceptions


class BaseResource(object):

    name = None
    path_pattern = None
    record_types = [
        'A', 'CNAME', 'MX', 'TXT', 'SRV', 'AAAA', 'SSHFP', 'PTR', 'ALIAS']

    def __init__(self, **kwargs):
        self.kwargs = kwargs

    @property
    def path(self):
        return self.path_pattern % self.kwargs

    def request(self, method, data=None, query_data=None):
        if data is not None:
            data = {self.name: data}
        return self.kwargs['request_func'](method, self.path, data, query_data)

    @classmethod
    def validate(cls, func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            if 'record_type' in kwargs and \
                    kwargs['record_type'].upper() not in cls.record_types:
                raise exceptions.RecordTypeError(
                    'Bad record: %s' % kwargs['record_type'])
            return func(*args, **kwargs)
        return wrapper


class BaseMember(BaseResource):

    def retrieve(self):
        response = self.request('get')
        if response.status == 200:
            return json.loads(response.content)
        if response.status == 404:
            raise exceptions.NotFoundError
        raise exceptions.UnknownError(response)

    @BaseResource.validate
    def update(self, **kwargs):
        response = self.request('put', kwargs)
        if response.status == 202:
            return
        if response.status == 422:
            raise exceptions.UnprocessableEntityError(
                json.loads(response.content))
        raise exceptions.UnknownError(response)

    def delete(self):
        response = self.request('delete')
        if response.status == 202:
            return
        if response.status == 409:
            raise exceptions.ConflictError
        raise exceptions.UnknownError(response)


class BaseCollection(BaseResource):

    member_class = None

    def __call__(self, id):
        kwargs = self.kwargs.copy()
        kwargs['id'] = id
        return self.member_class(**kwargs)

    def retrieve(self, **kwargs):
        response = self.request('get', query_data=kwargs)
        if response.status == 200:
            return json.loads(response.content)
        raise exceptions.UnknownError(response)

    @BaseResource.validate
    def create(self, **kwargs):
        response = self.request('post', kwargs)
        if response.status == 201:
            return json.loads(response.content)
        if response.status == 422:
            raise exceptions.UnprocessableEntityError(
                json.loads(response.content))
        raise exceptions.UnknownError(response)
